kotlinpoet/src/commonTest/kotlin/com/squareup/kotlinpoet/TypeSpecTest.kt

```diff
@@ -5778,6 +5779,385 @@ class TypeSpecTest {
+
+      protected class Taco {
+        val unspecifiedInts: IntArray
+
+        private val privateInts: IntArray
+
+        protected val protectedInts: IntArray
+
+        public val publicInts: IntArray
+
+        internal val internalInts: IntArray
+      }
+
+      """.trimIndent(),
+    )
+
+    val unspecifiedClass = TypeSpec.classBuilder("Taco")
+      .addProperty("unspecifiedInts", IntArray::class)
+      .addProperty("privateInts", IntArray::class, PRIVATE)
+      .addProperty("protectedInts", IntArray::class, PROTECTED)
+      .addProperty("publicInts", IntArray::class, PUBLIC)
+      .addProperty("internalInts", IntArray::class, INTERNAL)
+      .build()
+
+    //language=kotlin
+    assertThat(toString(unspecifiedClass)).isEqualTo(
+      """
+      package com.squareup.tacos
+
+      import kotlin.IntArray
+
+      public class Taco {
+        public val unspecifiedInts: IntArray
+
+        private val privateInts: IntArray
+
+        protected val protectedInts: IntArray
+
+        public val publicInts: IntArray
+
+        internal val internalInts: IntArray
+      }
+
+      """.trimIndent(),
+    )
+  }
+
+  @Test fun permissiveModifiersOmittedOnRestrictedVisibilityFunctionInClass() {
+    val unspecifiedVisibilityMethod = FunSpec.builder("unspecifiedStringMethod")
+      .returns(String::class)
+      .addStatement("return %S", "taco")
+      .build()
+
+    val privateVisibilityMethod = FunSpec.builder("privateStringMethod")
+      .returns(String::class)
+      .addModifiers(PRIVATE)
+      .addStatement("return %S", "taco")
+      .build()
+
+    val internalVisibilityMethod = FunSpec.builder("internalStringMethod")
+      .returns(String::class)
+      .addModifiers(INTERNAL)
+      .addStatement("return %S", "taco")
+      .build()
+
+    val protectedVisibilityMethod = FunSpec.builder("protectedStringMethod")
+      .returns(String::class)
+      .addModifiers(PROTECTED)
+      .addStatement("return %S", "taco")
+      .build()
+
+    val publicVisibilityMethod = FunSpec.builder("publicStringMethod")
+      .returns(String::class)
+      .addModifiers(PUBLIC)
+      .addStatement("return %S", "taco")
+      .build()
+
+    val privateClass = TypeSpec.classBuilder("Taco")
+      .addModifiers(PRIVATE)
+      .addFunction(unspecifiedVisibilityMethod)
+      .addFunction(privateVisibilityMethod)
+      .addFunction(internalVisibilityMethod)
+      .addFunction(protectedVisibilityMethod)
+      .addFunction(publicVisibilityMethod)
+      .build()
+
+    assertThat(toString(privateClass)).isEqualTo(
+      """
+      package com.squareup.tacos
+
+      import kotlin.String
+
+      private class Taco {
+        fun unspecifiedStringMethod(): String = "taco"
+
+        private fun privateStringMethod(): String = "taco"
+
+        internal fun internalStringMethod(): String = "taco"
+
+        protected fun protectedStringMethod(): String = "taco"
+
+        public fun publicStringMethod(): String = "taco"
```

Comment:
A bunch of these tests seem incorrect with the goal. In my original bug there's a private class showing that the public modifier isn't needed.

Reply:
I wanted to verify that if any access modifier was specified, it was respected and added to the property / function even if the contain class was private / internal / protected / unspecified / public. So I added a matrix of those 5 each containing a function across the same 5, with the main change in behavior being that where the class is private / internal / protected, unspecifiedStringMethod has no access modifier specified in the source code because no modifier was specified through kotlin poet. 

please let me know if i'm way off base!